Add 'watch' argument to status command

evidolob opened the pull request:
**Fixes:** Issue #N

**Relates to:** #2876

## Solution/Idea

This will display CPU and RAM load in real time.
To achieve this PR add websocket implementation and daemon send data through websocket `/status` endpoint.

It is use https://github.com/nhooyr/websocket as websocket implementation
and upgrade https://github.com/cheggaaa/pb to `v3` to render load bars in terminal.

Demo:

https://user-images.githubusercontent.com/929743/208875792-e665602a-548d-4f71-bda9-fd435bfc70ac.mov


## Testing

1. run any preset
2. call `crc status -w`
3.  look on output, it should contains bar for RAM and bars for CPU's load

evidolob committed:
Add '-w' argument to status command

Signed-off-by: Yevhen Vydolob <yvydolob@redhat.com>

anjannath commented on pkg/crc/api/websocket/status.go:
```diff
@@ -0,0 +1,52 @@
+}
+
+func (s StatusConnectionListener) onClientConnected(count int, sendData SendData) {
+	if count == 1 {
```
does it mean, only one ws `client` will be getting the status?

evidolob replied:
No, that mean that collection of CPU/RAM usage start only when first client connected, but all connected clients will receive info.

anjannath replied:
got it, and we don't want to start another go routine if one is already started and `count > 0` would start it for every client.

cfergeau replied:
Might be clearer as `Start`/`Stop` or similar naming? I find it a bit artificial to have a generic method which gets passed the number of connected client which then is a noop unless there's only 1 connected client.
(and this "count" can be misleading, there is no lock taken while it's called, so it means "around the time onClientConnected() was called, there was 1 client connected, but by the time you get to `if count==1`, any number of clients may have connected/disconnected")

anjannath commented on pkg/crc/api/websocket/handler.go:
```diff
@@ -0,0 +1,38 @@
+
+func NewEndpointHandler(data SendData) *EndpointHandler {
+	handler := &EndpointHandler{
+		listeners: make(map[ConnectionListener]struct{}),
```
maybe it would be better if we can avoid using these empty maps with only map keys but an empty `struct`?

anjannath commented on cmd/crc/cmd/daemon.go:
```diff
@@ -142,7 +143,9 @@ func run(configuration *types.Configuration) error {
-		mux.Handle("/api/", http.StripPrefix("/api", api.NewMux(config, newMachine(), logging.Memory, segmentClient)))
+		machineClient := newMachine()
+		mux.Handle("/api/", http.StripPrefix("/api", api.NewMux(config, machineClient, logging.Memory, segmentClient)))
+		mux.Handle("/socket/", http.StripPrefix("/socket", websocket.NewWebsocketServer(machineClient)))
```
wondering if this could've been part of the same `/api` path, something like `/api/statuswatch` ?

evidolob replied:
Hm, It may, let me try.

I prefer to separate HTTP API and WebSocket API, in URL. We may have only one websocket endpoint, and add some soer of PRC API or PubSub  API, and handle all events/query’s in single websocket connection. But this approach requires more complex client implementation.

anjannath replied:
okay, don't change things yet if it complicates the client implementations! i was just asking to know if it was possible or there was some limitations to using the same HTTP server for websocket as well as normal HTTP

evidolob commented:
/retest

cfergeau commented on cmd/crc/cmd/status.go:
```diff
@@ -48,11 +58,97 @@ type status struct {
+
+	ctx := context.Background()
+
+	con, _, err := websocket.Dial(ctx, "ws://unix/socket/status", &websocket.DialOptions{HTTPClient: &http.Client{
```
Have you considered moving this + the json unmarshalling to `daemonclient`?  It might fit better there.
`con` is usually spelt `conn` if it's a `net.Conn`

cfergeau commented on cmd/crc/cmd/status.go:
```diff
@@ -48,11 +58,97 @@ type status struct {
+		}
+
+		wmState := &types.WmStatusResult{}
+		err = json.Unmarshal(msg, wmState)
```
I think you could use `Read`  from https://pkg.go.dev/nhooyr.io/websocket/wsjson and skip the unmarshalling, but not sure it's going to bring us much.

cfergeau commented on cmd/crc/cmd/status.go:
```diff
@@ -48,11 +58,97 @@ type status struct {
+				return err
+			}
+		}
+		// '0' bar is RAM
```
I'd use multiple go return values instead of stuffing everything in an array:
```
func createBars(cpuUse []int64, writer io.Writer) (*pb.ProgressBar, []*pb.ProgressBar) {
   [...]
    return memoryBar, result
}
[...]
    ramBar, cpuBars := createBars(...)
[...]
```
You can also name the return values in the function prototype if you want to add a bit of self-documentation.

cfergeau commented on cmd/crc/cmd/status.go:
```diff
@@ -48,11 +58,97 @@ type status struct {
+		bar.SetTemplateString(tmpl)
+		bar.SetMaxWidth(150)
+
+		result[i+1] = bar
```
I'd just use 
```
    result := []*pb.ProgressBar{} // or var result []*pb.ProgressBar if this is favoured by goland/linting tools
    [...]
    result = append(result, memoryBar)
    for i := 0; i < len(cpuUse); i++ {
        [...]
        result = append(result, bar)
    }
```
This way you don't need to precompute the length of `result` or to think about the index you are putting the element at.

cfergeau commented on cmd/crc/cmd/status.go:
```diff
@@ -48,11 +58,97 @@ type status struct {
+	memoryBar.SetWriter(writer)
+	memoryBar.Set(pb.Bytes, true)
+	memoryBar.Set(pb.Static, true)
+	tmpl := `{{ red "RAM:" }} {{counters . }} {{percent .}} {{string . "my_green_string" | green}} {{ bar . "[" "\u2588" "\u2588" " " "]"}} `
```
On my test machine, the closing `]` for the memory bar was not aligned with the other ones , it was one char before.

evidolob replied:
Just fix that:
<img width="1070" alt="Screenshot 2023-01-23 at 16 46 27" src="https://user-images.githubusercontent.com/929743/214068794-f63adf4f-d8ca-485c-8663-15aa9617fc60.png">

cfergeau commented on pkg/crc/cluster/cluster.go:
```diff
@@ -104,6 +105,29 @@ func GetRAMUsage(sshRunner *ssh.Runner) (int64, int64, error) {
 
+func GetCPUUsage(sshRunner *ssh.Runner) ([]int64, error) {
+	cmd := "top -n1 -1 -b -w 79  | grep %Cpu | awk '{print $3}'"
+	out, _, err := sshRunner.Run(cmd)
```
I was a bit worried that the output would be locale-dependant, but this seems ok, if I set `LC_ALL`/`LANG`, manually start the daemon, `sshRunner.Run("printenv")` does not inherit the locale changes, which is good (floats in French locale are "6,2", not "6.2")

cfergeau commented on pkg/crc/cluster/cluster.go:
```diff
@@ -104,6 +105,29 @@ func GetRAMUsage(sshRunner *ssh.Runner) (int64, int64, error) {
+			return nil, err
+		}
+
+		result[i] = int64(math.Round(c))
```
similarly to one of my previous suggestions ,you could use `result = append(result, int64(...))` which removes the need for allocating the correct amount of memory and using the right index.

cfergeau commented on pkg/crc/cluster/cluster.go:
```diff
@@ -104,6 +105,29 @@ func GetRAMUsage(sshRunner *ssh.Runner) (int64, int64, error) {
 	return ramSize, ramUsage, nil
 }
 
+func GetCPUUsage(sshRunner *ssh.Runner) ([]int64, error) {
```
Can you add a short comment about the content of this `[]int64`?

cfergeau commented on pkg/crc/machine/client.go:
```diff
@@ -22,6 +22,7 @@ type Client interface {
 	PowerOff() error
 	Start(ctx context.Context, startConfig types.StartConfig) (*types.StartResult, error)
 	Status() (*types.ClusterStatusResult, error)
+	WmStatus() (*types.WmStatusResult, error)
```
Not sure what `Wm` stands for?

cfergeau reviewed (commented):
First round of comments, I haven't looked at `api/websocket` yet.

cfergeau commented on pkg/crc/api/websocket/endpoint.go:
```diff
@@ -0,0 +1,59 @@
+	addClient(client *wsClient)
+	deleteClient(client *wsClient)
+	setHandler(handler *EndpointHandler)
+	sendData([]byte)
```
This probably be replaced this with `io.Writer` which provides almost the same method, `Write(p []byte) (n int, err error)` and will be familiar to go developers.

cfergeau commented on pkg/crc/api/websocket/endpoint.go:
```diff
@@ -0,0 +1,59 @@
+
+type endpoint struct {
+	clients   map[*wsClient]struct{}
+	clientsMu sync.Mutex
```
Can you name this `clientsMutex` or `clientLock`?

cfergeau commented on pkg/crc/api/websocket/endpoint.go:
```diff
@@ -0,0 +1,59 @@
+	e.clients[client] = struct{}{}
+	e.clientsMu.Unlock()
+
+	e.handler.onClientConnected(len(e.clients))
```
`len(e.clients)` is unprotected by the mutex, my feeling is that this could lead to various (unlikely) races. Not sure something bad could happen apart from having 2 goroutines from `StatusConnectionListener.onClientConnected` running briefly at the same time before one of them is terminated.

cfergeau commented on pkg/crc/api/websocket/endpoint.go:
```diff
@@ -0,0 +1,59 @@
+	e.clientsMu.Lock()
+	defer e.clientsMu.Unlock()
+
+	for s := range e.clients {
```
`s`? Why not `client`? I get confused very quickly with one letter variable names :(

cfergeau commented on pkg/crc/api/websocket/endpoint.go:
```diff
@@ -0,0 +1,59 @@
+
+	for s := range e.clients {
+		select {
+		case s.msgs <- data:
```
This forwards data from the goroutine created in StatusConnectionListener.onClientConnected to the goroutines Server.subscribe runs in. This runs in the StatusConnectionListener.onClientConnected goroutine.
(It took me some code reading to get to this understanding which I hope is correct, it would not hurt to describe this somewhere imo, either in code comments or in commit logs).

I'm wondering if the client could embed its `*websocket.Conn` and we'd call directly the `writeTimeout` method  (or equivalent) from here to avoid an indirection? There is no code making use of the connection in the other goroutine.

If you prefer to keep things this way, I'd at least add a comment to explicitly say we forward data to the s.msgs channel, as it's a bit hidden in the `case`, and it's easy to think execution will fallthrough to the `default` case.

cfergeau commented on pkg/crc/api/websocket/server.go:
```diff
@@ -0,0 +1,126 @@
+	}
+
+	ws.serveMux.HandleFunc("/status", func(writer http.ResponseWriter, request *http.Request) {
+		ws.endpoints["/status"] = createStatusEndpoint(machine)
```
Shouldn't this be done outside of `HandleFunc`? Won't this create a new status endpoint each time we get a request on `/socket/status`?

cfergeau commented on pkg/crc/api/websocket/server.go:
```diff
@@ -0,0 +1,126 @@
+	statusEndpoint := NewEndpoint()
+	statusEndpointHandler := NewEndpointHandler(statusEndpoint.sendData)
+	statusEndpoint.setHandler(statusEndpointHandler)
+	statusEndpointHandler.addListener(NewStatusListener(machine))
```
A bit difficult for me to wrap my head around endpoint/handler/listener, it seems only 2 would be needed? Or is endpoint for sending data on the websocket, handler for client connection/disconnection, and listener to periodically get the status data?

evidolob replied:
>Or is endpoint for sending data on the websocket, handler for client connection/disconnection, and listener to periodically get the status data?

Yes, I try to create general websocket API, which will not bonded to `status` implementation.
In general we need to set up this once per endpoint(websocket url)

cfergeau commented on pkg/crc/api/websocket/status.go:
```diff
@@ -0,0 +1,52 @@
+						logging.Errorf("unexpected error during getting machine status: %v", err)
+					}
+
+					bytes, _ := json.Marshal(status)
```
error handling is missing

cfergeau commented on pkg/crc/api/websocket/server.go:
```diff
@@ -0,0 +1,126 @@
+
+	defer c.Close(websocket.StatusInternalError, "")
+
+	// TODO add reading messages form client
```
minor "form" typo

cfergeau commented on cmd/crc/cmd/status.go:
```diff
@@ -106,47 +107,45 @@ func runWatchStatus(writer io.Writer, client *daemonclient.Client, cacheDir stri
+	ramBar.SetTemplateString(tmpl)
 
-	for i := 0; i < len(cpuUse); i++ {
+	for i, _ := range cpuUse {
```
`, _` is not needed

cfergeau commented on pkg/crc/api/client/websocket_client.go:
```diff
@@ -0,0 +1,38 @@
+	}
+}
+
+func (c *WebSocketClient) Status(callback StatusCallback) error {
```
`func (c *WebSocketClient) Status(callback StatusCallback) (types.WmStatusResult, error)`  without the `for {}` loop would be more consistent with the other 'client' APIs  and would avoid the need for the callback. The `for{}` loop could stay in the `cmd/` package.

jsliacan commented:
/retest

cfergeau committed:
websocket: Use io.Writer interface

The SendData function type and the Endpoint.Write methods are very close
to what the io.Writer interface provides. Let's use it instead.

cfergeau committed:
websocket: Add a bit of API doc

cfergeau committed:
websocket: Make StatusConnectionListener generic

Only the call to get the status is status-specific, the rest of the code
marshalling the data, fetching it every few seconds, ... is applicable
to anything we want to send through websocket. This commit extracts the
'status' bit from the generic code, this will be useful if we add other
similar websocket calls.

cfergeau commented:
I've done a few commits in https://github.com/cfergeau/crc/commits/status-watch on top of your branch instead of proper review comments /o\ `websocket: Use io.Writer interface` `Make StatusConnectionListener generic` are mere suggestions, feel free to take them or ignore them as you prefer.
`Should invalid RAMUse/RamSize be set to 0 or -1?` is a question regarding error values in the loadresult structure, maybe 0 is correct, it was not clear to me. And the added API doc may need some fixing, this is my understanding of the split, but maybe some of the naming I use is incorrect.
Last commit is `s/Transport/transport` and should be fairly self-explanatory.

Apart from this, this looks good to me!

evidolob commented:
@cfergeau I'm OK with your changes, could you add them to this PR?

evidolob commented:
@cfergeau I’ve cherry-picked your commits

evidolob commented:
/retest

cfergeau reviewed (approved):
I squashed a number of commits together and force pushed them.  Let's get this in!

evidolob commented:
/test e2e-crc
